### PR TITLE
[`pylint`] Implement `invalid-length-returned` (`E0303`)

### PR DESCRIPTION
**Pull request by tibor-reiss**
Add pylint rule invalid-length-returned (PLE0303)

See https://github.com/astral-sh/ruff/issues/970 for rules

Test Plan: `cargo test`

TBD: from the description: "Strictly speaking `bool` is a subclass of `int`, thus returning `True`/`False` is valid. To be consistent with other rules (e.g. [PLE0305](https://github.com/astral-sh/ruff/pull/10962) invalid-index-returned), ruff will raise, compared to pylint which will not raise."

### CODE REVIEW

**Comment by tibor-reiss** [comment]
The tests run locally fine when using commit 670d66f as rebase, but there are failures when using the next commit, c221035. Was there some limit reached (rule is u16 in rule_set.rs / contains()) maybe?

**Comment by charliermarsh** [comment]
Try bumping the value in `const RULESET_SIZE: usize = 13;`

**Comment by tibor-reiss** [comment]
> Try bumping the value in `const RULESET_SIZE: usize = 13;`

@charliermarsh Is the performance drop of 4% also related to this please? The newly introduced rule is quite lightweight.

File: crates/ruff_linter/src/checkers/ast/analyze/statement.rs [hunk 1]
```diff
@@ -97,6 +97,9 @@ pub(crate) fn statement(stmt: &Stmt, checker: &mut Checker) {
             if checker.enabled(Rule::InvalidBoolReturnType) {
                 pylint::rules::invalid_bool_return(checker, name, body);
             }
+            if checker.enabled(Rule::InvalidLengthReturnType) {
+                pylint::rules::invalid_length_return(checker, name, body);
```
Comment (MichaReiser) [inline crates/ruff_linter/src/checkers/ast/analyze/statement.rs:101]: Nit: You can pass the entire function definition instead of just passing the name and body
```suggestion
                pylint::rules::invalid_length_return(checker, function_def);
```

File: crates/ruff_linter/src/rules/pylint/rules/invalid_length_return.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    if name != "__len__" {
+        return;
+    }
+
+    if !checker.semantic().current_scope().kind.is_class() {
+        return;
+    }
+
+    let returns = {
+        let mut visitor = ReturnStatementVisitor::default();
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/pylint/rules/invalid_length_return.rs:58]: This is not really related to your change but be should change the `ReturnStatementVisitor` to not visit nested bodies. 

```
class Test:
    def __len__(self):
        def nested():
            return "abcd" # we don't want to flag this statmenet

        return 40     
```

You can do this by changing `visit_stmt` and adding explicit branches for `FunctionDef` and `StatementDef`

```rust
    fn visit_stmt(&mut self, stmt: &'a Stmt) {
        match stmt {
            Stmt::FunctionDef(_) | Stmt::ClassDef(_) => {
                // Don't recurse.
            }
            Stmt::Return(stmt) => self.returns.push(stmt),
            Stmt::FunctionDef(_) | Stmt::ClassDef(_) => {},
            _ => crate::visitor::walk_stmt(self, stmt),
        }
    }
```

I think it would be good to add a test for it.

File: crates/ruff_linter/src/rules/pylint/rules/invalid_length_return.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                .push(Diagnostic::new(InvalidLengthReturnType, stmt.range()));
+        }
+    }
+}
+
+fn is_negative_integer(value: &Expr) -> bool {
+    let Expr::UnaryOp(ast::ExprUnaryOp { op, .. }) = value else {
+        return false;
+    };
+    matches!(op, ast::UnaryOp::USub)
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/pylint/rules/invalid_length_return.rs:88]: Nit: 
```suggestion
        matches!(
        value,
        Expr::UnaryOp(ast::ExprUnaryOp {
            op: ast::UnaryOp::USub,
            ..
        })
    )
```

File: crates/ruff_linter/src/rules/pylint/rules/invalid_length_return.rs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    };
+
+    for stmt in returns {
+        if let Some(value) = stmt.value.as_deref() {
+            if !matches!(
+                ResolvedPythonType::from(value),
+                ResolvedPythonType::Unknown
+                    | ResolvedPythonType::Atom(PythonType::Number(NumberLike::Integer))
+            ) || is_negative_integer(value)
+            {
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/pylint/rules/invalid_length_return.rs:70]: It could make sense to change the order of the checks because `is_negative_integer` is relatively simple compared to `ResolvedPythonType::from`
```suggestion
            if is_negative_integer(value) ||
             !matches!(
                ResolvedPythonType::from(value),
                ResolvedPythonType::Unknown
                    | ResolvedPythonType::Atom(PythonType::Number(NumberLike::Integer))
            )
            {
```

**Review by MichaReiser — APPROVED** [review]

**Comment by MichaReiser** [comment]
@tibor-reiss I bumped the index (and fixed the perf regression). You can rebase and we should then see the impact of your rule.

File: crates/ruff_linter/src/rules/pylint/rules/invalid_length_return.rs (hunk 3 above)
Comment (tibor-reiss) [reply crates/ruff_linter/src/rules/pylint/rules/invalid_length_return.rs:88]: Thanks, done!

File: crates/ruff_linter/src/rules/pylint/rules/invalid_length_return.rs (hunk 4 above)
Comment (tibor-reiss) [reply crates/ruff_linter/src/rules/pylint/rules/invalid_length_return.rs:70]: TLDR: changed.
I was not sure about which should be first because I would expect that there are more cases where the return type is wrong, compared to where the type is ok (int) but the sign not.
